### ISSUE TITLE
Hazelcast ITopic - produce event to all -other- members/consumers in a cluster

### ISSUE THREAD

**Issue by eldarj** [hazelcast/hazelcast#16974]
Currently, e.g. if having a single topic in a 3-member-cluster, and a single producer/single consumer per member - there's actually no way to avoid consuming/reacting to self-sent events.

E.g. this is what happens in a cluster with 3 members; M1, M2, M3; where all members can produce and consume from same topic:
- M1 produces eventA to some-topic
- M1, M2, M3 consume eventA from some-topic
- M2 produces to eventB some-topic
- M1, M2, M3 consume eventB from some-topic

Suggested/needed behaviour:
- M1 produces eventA to some-topic
- M2, M3 consume eventA from some-topic
- M2 produces to eventB some-topic
- M1, M3 consume eventB from some-topic

This can be avoided with either defining multiple topics and implementing some routing, or implementing manual checks and an envelope-like container (with some kind of producerId) for produced/consumed events. These approaches cause overhead I believe.

Would it be sensible to provide some kind of out-of-the-box configuration/policy that would let us choose whether produced events from a specific cluster member should end up for consumption, on that same member?

**Comment by mmedenjak** [comment]
Yes, you're right that it would be possible to implement this but we're not considering it at the moment. For now, you can, as you say, add some kind of producer ID to each message, and filter out subscribed messages based on the current node ID.

For now, I'm putting this issue in backlog so we see if there is more interest.

**Comment by eldarj** [comment]
How does community contribution work? I'm more than happy to work on this, if the enhancement makes sense and would be useful.

**Comment by mmedenjak** [comment]
Hi @eldarj ! You can check out these links:
https://github.com/hazelcast/hazelcast#contributing
https://github.com/hazelcast/hazelcast/blob/master/CONTRIBUTING.md

In short, you fork the repository, create a branch probably based off of master (if you want to work with the latest code) and once you're ready, submit a pull request against our master. We'll pick it up from there, review it and possibly get it merged.

We can also provide you with some initial pointers for the implementation if you need them.
